developers/weaviate/client-libraries/typescript.mdx

```diff
@@ -0,0 +1,162 @@
+  authClientSecret: new weaviate.AuthUserPasswordCredentials({
+    username: "user123",
+    password: "password",
+    scopes: ["offline_access"]  // optional, depends on the configuration of your identity provider (not required with WCS)
+  })
+});
+```
+
+### Client Credentials flow
+
+import ClientAuthFlowClientCredentials from '/developers/weaviate/client-libraries/_components/client.auth.flow.client.credentials.mdx'
+
+<ClientAuthFlowClientCredentials />
+
+```ts
+const client: WeaviateClient = weaviate.client({
+  scheme: "https",
+  host: "some-endpoint.weaviate.network",
+  authClientSecret: new weaviate.AuthClientCredentials({
+    clientSecret: "supersecuresecret",
+    scopes: ["scope1", "scope2"]  // optional, depends on the configuration of your identity provider
+  })
+});
+```
+
+### Refresh Token flow
+
+import ClientAuthBearerToken from '/developers/weaviate/client-libraries/_components/client.auth.bearer.token.mdx'
+
+<ClientAuthBearerToken />
+
+```ts
+const client: WeaviateClient = weaviate.client({
+  scheme: "https",
+  host: "some-endpoint.weaviate.network",
+  authClientSecret: new weaviate.AuthAccessTokenCredentials({
+    accessToken: "abcd1234",
+    expiresIn: 900,
+    refreshToken: "efgh5678"
+  })
+});
+```
+
+## Custom headers
+
+You can pass custom headers to the client, which are added at initialization:
+
+```ts
+const client: WeaviateClient = weaviate.client({
+  scheme: "https",
+  host: "some-endpoint.weaviate.network",
+  headers: {headerName: 'HeaderValue'}
+});
+```
+
+These headers will then be included in every request that the client makes.
+
+## References
+
+:::note JS → TS migration
+We are in the process of migrating our JavaScript code examples to TypeScript as necessary. Please be patient.
+:::
+
+Our [RESTful endpoints](../api/rest/index.md) and [GraphQL functions](../api/graphql/index.md) covered by the TypeScript client currently have JavaScript examples in the code blocks.
+
+## Design
+
+### Builder pattern
+
+The TypeScript client is designed following the [Builder pattern](https://en.wikipedia.org/wiki/Builder_pattern). The pattern is used to build complex query objects. This means that calls (for example to retrieve data with a RESTful GET request, or using a more complex GraphQL query) are built with chained objects to reduce complexity. Some builder objects are optional, others are required to perform specific functions. Builder examples can be found in the [RESTful API reference pages](../api/rest/index.md) and the [GraphQL reference pages](../api/graphql/index.md).
+
+The code snippet at the top of this page shows a simple query corresponding to the RESTful request `GET /v1/schema`. The client is initialized by requiring the package and connecting to a local instance. Then, a query is constructed by getting the `.schema` with `.getter()`. The query will be sent with the `.do()` call. `do()` is required for every function you want to build and execute.
+
+### General notes
```

Comment:
Very nice info here 👍

Reply:
Nice! That's probably Sebastian or whomever wrote the JS docs. I can't take any credit haha.

Reply:
I remember writing the "`do()` is required for every function you want to build and execute" part :)